File: JavaScript/MaxPriorityQueue.js
import { maxHeapify } from './HeapSort';

export default class MaxPriorityQueue {
  constructor(heap = []) {
    this.heap = heap;
    this.heapSize = 0;
  }

  parent(i) {
    return Math.floor(i / 2);
  }

  maximum() {
    return this.heap[0];
  }

  extractMax() {
    if (this.heapSize < 1) {
      throw 'Heap Underflow'
    }
    const n = this.heapSize - 1;
    const maxElement = this.maximum();
    this.heap[0] = this.heap[n];
    this.heap.pop();
    this.heapSize -= 1;
    maxHeapify(this.heap, this.heapSize, 0);
    return maxElement;
  }

  increaseKey(i, key) {
    if (key < this.heap[i]) {
      throw 'New key is smaller than current key'
    }
    this.heap[i] = key;
    while (this.heap[this.parent(i)] < this.heap[i]) {
      [this.heap[i], this.heap[this.parent(i)]] = [this.heap[this.parent(i)], this.heap[i]];
      i = this.parent(i);
    }
  }

  insert(key) {
    this.heapSize += 1;
    this.heap.push(-1);
    this.increaseKey(this.heapSize - 1, key);
  }

  build(arr) {
    this.heapSize = 0;
    this.heap = []
    arr.forEach((val) => {
      this.insert(val);
    });
  }
}

File: Python/MaxPriorityQueue.py
from Python.HeapSort import maxHeapify

class MaxPriorityQueue:
  def __init__(self, heap=[]):
    self.heap = heap
    self.heapSize = len(heap)

  def parent(self, i):
    return i // 2

  def maximum(self):
    return self.heap[0]

  def extractMax(self):
    if(self.heapSize < 1):
      raise Exception('Heap Underflow')
    n = self.heapSize - 1
    maxElement = self.maximum()
    self.heap[0] = self.heap[n]
    self.heap.pop() # Remove the element from the heap now that it is extracted
    self.heapSize = self.heapSize - 1
    maxHeapify(self.heap, self.heapSize, 0)
    return maxElement

  def increaseKey(self, i, key):
    if key < self.heap[i]:
      raise Exception('New key is smaller than current key.')
    self.heap[i] = key
    while self.heap[self.parent(i)] < self.heap[i]:
      self.heap[i], self.heap[self.parent(i)] = self.heap[self.parent(i)], self.heap[i]
      i = self.parent(i)

  def insert(self, key):
    self.heapSize = self.heapSize + 1
    self.heap.append(-1)
    self.increaseKey(self.heapSize - 1, key)

  def build(self, arr):
    self.heapSize = 0
    self.heap =[] # Reset when you're building from an array
    for i in range(len(arr)):
      self.insert(arr[i])
